cmake/opentitan.cmake

```diff
@@ -0,0 +1,104 @@
+# Copyright 2024 ETH Zurich and University of Bologna.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Viviane Potocnik <vivianep@iis.ee.ethz.ch>
+
+include(ExternalProject)
+
+set(OPENTITAN_DIR ${CMAKE_CURRENT_BINARY_DIR}/../drivers/opentitan)
+
+if(EXISTS ${OPENTITAN_DIR})
+  file(REMOVE_RECURSE ${OPENTITAN_DIR})
+endif()
+
+set(SPARSE_CHECKOUT_FILE ${CMAKE_CURRENT_BINARY_DIR}/sparse-checkout)
+
+set(SPARSE_PATTERNS
+  "util/*"
+  "hw/ip/*"
+  "sw/device/lib/base/*"
+  "sw/device/lib/base/internal/*"
+  "sw/device/lib/dif/*"
+  "sw/device/lib/dif/autogen/*"
+)
+
+file(WRITE ${SPARSE_CHECKOUT_FILE} "")
+foreach(PATTERN ${SPARSE_PATTERNS})
+  file(APPEND ${SPARSE_CHECKOUT_FILE} "${PATTERN}\n")
+endforeach()
+
+# Add OpenTitan as an external project
+ExternalProject_Add(
+  opentitan
+  DOWNLOAD_COMMAND
+    git clone --no-checkout https://github.com/lowRISC/opentitan.git ${OPENTITAN_DIR}
+    COMMAND git -C ${OPENTITAN_DIR} config core.sparseCheckout true
+    COMMAND ${CMAKE_COMMAND} -E copy ${SPARSE_CHECKOUT_FILE} ${OPENTITAN_DIR}/.git/info/sparse-checkout
+    COMMAND git -C ${OPENTITAN_DIR} checkout master
+  UPDATE_COMMAND ""  # Disable updates to avoid unexpected changes
+  CONFIGURE_COMMAND ""  # No configure step needed
+  BUILD_COMMAND ""  # No build step needed
+  INSTALL_COMMAND ""  # No install step needed
+  SOURCE_DIR ${OPENTITAN_DIR}
+)
+
+message(STATUS "OpenTitan added as an external project.")
+
+set(OPENTITAN_SW_DIR ${OPENTITAN_DIR}/sw)
+set(OPENTITAN_HW_IP_DIR ${OPENTITAN_DIR}/hw/ip)
+set(OPENTITAN_UTIL_DIR ${OPENTITAN_DIR}/util)
+
+set(REGTOOL_PY ${OPENTITAN_UTIL_DIR}/regtool.py)
+set(UART_HJSON ${OPENTITAN_HW_IP_DIR}/uart/data/uart.hjson)
+set(UART_REGS_H ${OPENTITAN_SW_DIR}/device/lib/dif/autogen/uart_regs.h)
+
+add_custom_command(
+  OUTPUT ${UART_REGS_H}
+  COMMAND ${Python3_EXECUTABLE} ${REGTOOL_PY} ${UART_HJSON} -D > ${UART_REGS_H}
+  DEPENDS ${REGTOOL_PY} ${UART_HJSON}
+  COMMENT "Generating uart_regs.h"
+)
+
+add_custom_target(
+  generate_uart_regs ALL
+  DEPENDS ${UART_REGS_H}
+)
```

Comment:
Can we externalize this? I would do this in a setup procedure, not the main build flow.

Reply:
Thanks for the feedback! I see the concern about keeping the `uart_regs.h` generation within the main build flow, and I’d like to clarify why this step is currently integrated and how externalizing it might affect the workflow.

The `uart_regs.h` is necessary for building the `opentitan_lib`. Without it, the library cannot be compiled. Externalizing this process entirely (e.g., as a manual pre-build step) would introduce additional steps for developers, like running a separate script before invoking the main build. Or maybe I am missing the point of your comment 😅 

If you’re suggesting a more decoupled approach (e.g., fully externalizing the process to a pre-build script or tool), let me know what you envision. 

- Would you prefer having a dedicated `setup.sh` or similar to prepare the repository before building?
- Or is the suggestion to make the build system more modular while retaining the automation? 

Let me know what you think @Scheremo .

Reply:
Seems like a perfect use case for a makefile, imo. Redownloading the entire repo and building everything from scratch requires an internet connection, but it's not really necessary. I'd add a very simple setup step (note: makefile, not shell script) if necessary to generate these headers. Another fairly reasonable option is to check in generated headers in a patch.